File: 21-merge-two-sorted-lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        head = dummy = ListNode()
        # return head, iterate dummy

        # While both ptrs are not pointing at null (still going thru list)
        while list1 and list2: 
        # if one of the lists finishes, the other is already sorted, so we can stop
            if list1.val < list2.val:
                dummy.next = list1
                list1 = list1.next 
            else: 
                dummy.next = list2 # set the next
                list2 = list2.next
            dummy = dummy.next # actually go to the next (iterate)
         
        # determine which is list is left
        # then list1 or list2 ptr should point at the rest of the list
        # so just set dummy.next as equal to list1 or list2
        if list1:
            dummy.next = list1
        else:
            dummy.next = list2
        
        return head.next

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
    //while list1 and list 2 both still exist
    //when the loop breaks, only one of them will be left
    // add that ptr to that to the list

    struct ListNode* head = malloc(sizeof(struct ListNode)); 
    // This will stay in place

    head->val = 0;
    head->next = NULL;

    struct ListNode* list = head;

    while(list1 && list2) {
        // if list1.val >= list2.val, 
        // list.next = list2
        // list2 = list2.next

        // else list.next = list1
        // list1 = list1.next
        if (list1->val >= list2->val){
            list->next = list2;
            list2 = list2->next;
        } else {
            list->next = list1;
            list1 = list1->next;
        }

        list = list->next;
    }

    // what if one of the arrays still has nodes left?
    if (list1){
        list->next = list1;
    } 
    if (list2){
        list->next = list2;
    }

    return head->next; // list is currently the node before the beginning.
}